Crypto wrapper - implements tag based en/decryption utils

protich opened the pull request:
First stab at portable/interchangeable tag based encryption/decryption utils

Cryptos supported so far are...
- mcrypt
- openssl
- phpseclib Crypt_AES

Addresses issue #627

protich committed:
Crypto wrapper - implements tag based en/decryption

protich committed:
Crypto wrapper - implements tag based en/decryption

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+
+    function decrypt($ciphertext, $key, $skey='encryption') {
+
+        if(!preg_match('/^\\$(\d)\\$(.*)/', $ciphertext, $result))
```
This might be cleaner and a bit faster

``` php
if (substr($ciphertext, 0, 1) != '$')
    return false

list (, $crypt, $ciphertext) = explode('$', $ciphertext, 3);
```

_`$crypt` is checked on line 56 to ensure it is valid_

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+
+         // Remove the padding.
+         $pad = ord($plaintext[strlen($plaintext) -1]);
+         $plaintext = substr($plaintext, 0, strlen($plaintext) - $pad);
```
This approach rocks. I would have done it like

```
$<cid>$>cyphertext>
```

where `$<cyphertext>` is `$<len>$<cleartext>`

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+            return false;
+
+        $crypt = $result[1]; //Crypt used..
+        $ciphertext = $result[2]; //encrypted  input.
```
``` php
$ciphertext = base64_decode($result[2]); //encrypted  input.
```

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+            $cryptos[CRYPT_OPENSSL] = new CryptoOpenSSL(CRYPT_OPENSSL);
+
+        if(defined('CRYPT_PHPSECLIB') && class_exists('CryptoPHPSecLib'))
+            $cryptos[CRYPT_PHPSECLIB] = new CryptoPHPSecLib(CRYPT_PHPSECLIB);
```
``` php
static $cryptos = false;
if ($cryptos === false) {
    $cryptos = array();
    if(defined('CRYPT_MCRYPT') && class_exists('CryptoMcrypt'))
        $cryptos[CRYPT_MCRYPT] = new CryptoMcrypt(CRYPT_MCRYPT);
    ...
}

return $cryptos;
```

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+        $this->tag_number = $tag;
+    }
+
+    function getTagNumber() {
```
Add an `exists` method _abstract_ to return false if undefined in child classes

``` php
function exists() {
    return false;
}
```

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+        //cycle thro' available cryptos
+        foreach($cryptos as $crypto) {
+            if(is_callable(array($crypto, 'exists'))
+                    && call_user_func(array($crypto, 'exists')))
```
With the abstract `exists` method, it should always be _callable_

``` php
if (call_user_func(array($crypto, 'exists')))
```

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+
+    /* Hash string  - sha1 is used by default */
+    function hash($string, $key) {
+        $hash = new Crypt_Hash();
```
Specify SHA256 or SHA512 to get sufficiently long keys

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+    function hash($string, $key) {
+        $hash = new Crypt_Hash();
+        $hash->setKey($key);
+        return base64_encode($hash->hash($string));
```
Don't convert to text here. Ensure the crypto hash algorithm returns binary data and return it directly

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+
+        $hash = Crypto::hash($this->getMasterKey().md5($this->getSubKey()), $seed);
+        return $seed? substr($hash, 0, strlen($seed)) : $hash;
+    }
```
Ensure that the length of the key returned is guaranteed

``` php
function getKeyHash($seed, $length=32) { // 256-bit if unspecified
    $hash = Crypto::hash($this->getMasterKey().md5($this->getSubKey()), $seed);
    return substr($hash, 0, $length);
}
```

greezybacon commented on include/class.crypto.php:
```diff
@@ -0,0 +1,356 @@
+        $td = mcrypt_module_open($cipher['name'], '', $cipher['mode'], '');
+        $keysize = mcrypt_enc_get_key_size($td);
+        $ivsize = mcrypt_enc_get_iv_size($td);
+        $iv = mcrypt_create_iv($ivsize, MCRYPT_DEV_RANDOM); //XXX: Windows??
```
Use 

``` php
$iv = Crypto::randcode($ivsize);
```

which should be more consistent between *nix and Windows®

Commit:
Add documentation for the crypto process

protich committed:
Merge remote branch 'jared/feature/crypto' into feature/crypto
Refactor logic'

Conflicts:
	include/class.crypto.php

protich committed:
Add nolint to obvious false positives

protich committed:
Refactor PHPSECLIB Crypto
Add test function - basic testing (will be replaced by real test suit)

Commit:
Add a test suite for the crypto libraries

Only use 8 bytes (64-bits) for the IV data (salt) for the encryption keys,
except for Mcrypt, which requires you to query the specific length and use
it specifically.

protich committed:
Move getCipher to base class
Add cipher validator - checks to make sure the cipher is valid & available.

protich committed:
Merge remote branch 'jared/feature/crypto' into feature/crypto

protich committed:
Revert back to using cipher specific iv len

protich committed:
Remove temp. password

Commit:
Revisit migration tasks

* Move migration scripted tasks to the stream folder naming them .task.php.
* Add a MigrationTask abstract class to server as the base of migration
  tasks.
* Move the migration tasks from class.upgrader to the new task files

protich committed:
Merge remote branch 'jared/feature/crypto' into feature/crypto-upgrade

greezybacon commented:
You'll need to edit `setup/test/tests/test.crypto.php` and change the test values for openssl and phpseclib to include the 16-byte iv values

```
JDEkRiLtWBgRN68kJjp4jsM6xKJY+XFYwMeaQIHJXKW8v3fEZzs3gCq3hKevgvAjvdgwx5ZUYLFPsFehLtkzAw8IlQ==
```

for openssl and

```
JDEkvH/es2Drdsmc8pU2UBnBxhiPavtvst2Sl9jOYVXTRjHsgPmv8+8mIgwnA1nQ6EI2AoTq2gMZtoBoqK3Mzpw8IQ==
```

for phpseclib

protich committed:
Change email en/decryption to use crypto class
Drop old mcrypt class

protich committed:
Use 16-bytes iv.

protich committed:
Upgrade patch for 16fcef4a13d6475a5f8bfef462b548e2

protich committed:
Fix regression on upgrader
